cpp/src/arrow/compute/compute-test.cc

```diff
@@ -236,6 +236,65 @@ TEST_F(TestCast, ToIntDowncastSafe) {
   // underflow
   vector<int32_t> v6 = {0, 1000, 2000, -70000, 0};
   CheckFails<Int32Type>(int32(), v6, is_valid, int16(), options);
+
+  vector<int32_t> v7 = {0, 1000, 2000, -70000, 0};
+  CheckFails<Int32Type>(int32(), v7, is_valid, uint8(), options);
+}
+
+TEST_F(TestCast, IntegerSignedToUnsigned) {
+  CastOptions options;
+  options.allow_int_overflow = false;
+
+  vector<bool> is_valid = {true, false, true, true, true};
+
+  vector<int32_t> v1 = {INT32_MIN, 100, -1, 0, INT32_MAX};
+  // Same width
+  CheckFails<Int32Type>(int32(), v1, is_valid, uint32(), options);
+  // Wider
+  CheckFails<Int32Type>(int32(), v1, is_valid, uint64(), options);
+  // Narrower
+  CheckFails<Int32Type>(int32(), v1, is_valid, uint16(), options);
```

Comment:
In this case, the negative value `INT32_MIN` would trigger failure before the INT32_MAX overflow. Maybe add an additional test case wher all the values are positive, but there is a large-value overflow? To test for edge cases, you might want to test that `std::numeric_limits<uint16_t>::max()` is ok but  `std::numeric_limits<uint16_t>::max() + 1` is not